src/main/java/org/pkcs11/jacknji11/jna/JNA.java

```diff
@@ -52,48 +52,60 @@
             ? ULong.ULongSize.ULONG4 : ULong.ULongSize.ULONG8;
     }
 
+    private JNANativeI jnaNative = null;
+    
+    public JNA(){
+        this("cryptoki");
+    }
+    
+    public JNA(String customLibrary) {
+        jnaNative = (JNANativeI) com.sun.jna.Native.loadLibrary(customLibrary, JNANativeI.class);
+    }
+
     public long C_Initialize(CK_C_INITIALIZE_ARGS pInitArgs) {
-        return JNANative.C_Initialize(new JNA_CK_C_INITIALIZE_ARGS(pInitArgs));
+        if(pInitArgs.createMutex == null && pInitArgs.destroyMutex == null && pInitArgs.lockMutex == null && pInitArgs.unlockMutex == null)
```

Comment:
This null checks does not seem to have anything to do with the introduction of JNANativeI or?
Is it maybe a different solution for the issue in #5? Just curios and wanted to check so that it was not added by mistake.

Reply:
I think we need to revert the change in C_Initialize as unless there is a mutex provided the code misses to pass on the flags in pInitArgs. Specifically, the flag CK_C_INITIALIZE_ARGS.CKF_OS_LOCKING_OK will not be passed on which results in the process crashing in unspecified ways (malloc, free, sigsegv etc.) if running with multiple threads.
The issue has been confirmed with SoftHSMv2 at least as soon as you run more than 1 thread. Reverting this change in C_Initialize fixes the issue.